src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncExceptionHandlerRewriter.cs

```diff
@@ -888,14 +888,17 @@ public AwaitFinallyFrame()
                 // root frame
             }
 
-            public AwaitFinallyFrame(AwaitFinallyFrame parent, HashSet<LabelSymbol> labelsOpt, TryStatementSyntax tryStatementSyntax)
+            public AwaitFinallyFrame(AwaitFinallyFrame parent, HashSet<LabelSymbol> labelsOpt, StatementSyntax statementSyntax)
             {
                 Debug.Assert(parent != null);
-                Debug.Assert(tryStatementSyntax != null);
+                Debug.Assert(statementSyntax != null);
+
+                Debug.Assert(statementSyntax.Kind() == SyntaxKind.TryStatement || 
+                    (statementSyntax.Kind() == SyntaxKind.UsingStatement && ((UsingStatementSyntax)statementSyntax).AwaitKeyword != default));
```

Comment:
>statementSyntax.Kind() == SyntaxKind.UsingStatement [](start = 21, length = 51)

Is using statement syntax is sufficient differentiator? We might have multiple nested try/finally blocks originated from single using statement. Do we have tests for a scenario like that?

Reply:
I'm not sure I understood. The syntax is not used as a differentiator here. We just assert that awaits-in-finally originated from either a try-statement or an async-using syntax.

I added some tests with multiple resources and try/catch.